### ISSUE TITLE
Incorrect FnAttr applied to BLAS calls

### ISSUE THREAD

**Issue by ZuseZ4** [EnzymeAD/Enzyme#804]
For dot we generate one derivcall (axpy based), which we re-use to calculate dx and dy.
We do add ArgumentParameters to the call beforehead, however do not account for the change of Argument orders between both calls, which imho leads to the incorrect application of them. https://github.com/EnzymeAD/Enzyme/blob/04caf58bfa58ba078551187b965d0be17eff2782/enzyme/Enzyme/AdjointGenerator.h#L5991
So for correctness I would need to either create two axpy calls here or strip paramAttrs at location 2 and 4 and re-add them based on the updated (switched) conditions at line 6003, right?

**Comment by wsmoses** [comment]
I don't understand what you mean here. The parameters are attached to the function, which is independent of however it is called?

**Comment by ZuseZ4** [comment]
```.cpp
            if (call.getArgOperand(1)->getType()->isPointerTy()) {
              F->addParamAttr(2, Attribute::ReadOnly);
              F->addParamAttr(2, Attribute::NoCapture);
            }
            if (call.getArgOperand(3)->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }

                if (!gutils->isConstantValue(call.getArgOperand(3))) {
                  Value *args1[6] = {count, dif, xdata, xinc, dy, trueYinc};
                  Builder2.CreateCall(
                      derivcall, args1,
                      gutils->getInvertedBundles(
                          &call,
                          {ValueType::None,
                           xcache ? ValueType::None : ValueType::Primal,
                           ValueType::None, ValueType::Shadow, ValueType::None},
                          Builder2, /*lookup*/ true));
                }
                /// (...)
                if (!gutils->isConstantValue(call.getArgOperand(1))) {
                  Value *args2[6] = {count, dif, ydata, yinc, dx, trueXinc};
                  Builder2.CreateCall(
                      derivcall, args2,
                      gutils->getInvertedBundles(
                          &call,
                          {ValueType::None, ValueType::Shadow, ValueType::None,
                           ycache ? ValueType::None : ValueType::Primal,
                           ValueType::None},
                          Builder2, /*lookup*/ true));
                }            
```           
In the first call we have 
`Value *args1[6] = {count, dif, xdata, xinc, dy, trueYinc};`
Therefore dy will be modified.
This way it makes sense that we have
```.cpp
            if (call.getArgOperand(3)->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
```
since this will set NoCapture based on whether y (primary arg 3) is passed as pointer.
Nowe if we look at the second case we have
`Value *args2[6] = {count, dif, ydata, yinc, dx, trueXinc};`
Therefore dx will be modified. 
Shouldn't we in this case replace the condition to be
```.cpp
            if (call.getArgOperand(1)->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
```

**Comment by wsmoses** [comment]
What's being set here is the parameters of the function daxpy(int count, double v0, double* v1, int v1_inc, double* v2, int v2_inc);

The definition of daxpy always only reads from the 2nd arg (in this declaration v1).

In the first time we call daxpy we call it with:
```
daxpy(count, dif, xdata, xinc, dy, trueYinc);
```

Indeed the 2nd argument to this call (xdata) is only read.


In the first time we call daxpy we call it with:
```
daxpy(count, dif, ydata, yinc, dx, trueXinc);
```

Indeed the 2nd argument to this call (ydata) is only read.

**Comment by wsmoses** [comment]
You can confirm this behavior by looking at the definition of daxpy here: https://netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga8f99d6a644d3396aa32db472e0cfc91c.html. Which indeed only ever reads  from the 2nd argument.

**Comment by ZuseZ4** [comment]
Yes, I'm aware which part is from dot and which one from axpy and what their formula is.
So let me rephrase my example. We currently run the following code:
```.cpp
            if (inputArg_y->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
            Value *args1[6] = {count, dif, xdata, xinc, dy, trueYinc};
            daxpy(args1)
            Value *args2[6] = {count, dif, ydata, yinc, dx, trueXinc};                            
            daxpy(args2);
```
I think, that we instead would need to have the following code to be correct(same mod required for arg 2):
```.cpp
            if (inputArg_y->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
            Value *args1[6] = {count, dif, xdata, xinc, dy, trueYinc};
            daxpy(args1)
            F->removeParamAttrs(4);
            if (inputArg_x->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
            Value *args2[6] = {count, dif, ydata, yinc, dx, trueXinc};                            
            daxpy(args2);
```

**Comment by wsmoses** [comment]
No we do not need to do that, nor would it be semantically meaningful.

We are changing the attributes of the function daxpy being called, not the parameters we recieve from ddot. 

We can unconditionally apply the readonly/nocapture to F, if the calling convention is correct and it takes a pointer rather than an integer.

Julia by default emits the declaration of functions to take an i64 instead of a pointer (like a double*) and it is illegal to mark readonly/nocapture on an integer argument.

In other words, if the type signature of daxpy is as follows, we should apply attributes unconditionally:
```
daxpy(int, double, double*, int, double*, int);
```

However, if the daxpy already exists using the Julia calling convention it will appear like this (assuming the size of a pointer is the size of an int64), LLVM won't permit us to mark the integer argument as read only.

```
daxpy(int, double, int64, int, int64, int);
```

An equivalent but perhaps more clear way to have written this would be:
```
            if (F->getArgument(2)->getType()->isPointerTy()) {
              F->addParamAttr(2, Attribute::ReadOnly);
              F->addParamAttr(2, Attribute::NoCapture);
           }
            if (F->getArgument(4)->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
```

However, since we just defined the arg type of 2 to be the same as call.getArgOperand(...).getType(), that's used instead.

**Comment by ZuseZ4** [comment]
My point is that `dot(x,y) = dot(y,x)`, however our implementation does not account for that, due to how we set our ParamAttr for daxpy args 2 and 4.

**Comment by ZuseZ4** [comment]
Ok, I assumed that it was to different calling conventions, I was just confused since the conditions in the existing code were oddly specific. 
```.cpp
            if (F->getArgument(2)->getType()->isPointerTy()) {
              F->addParamAttr(2, Attribute::ReadOnly);
              F->addParamAttr(2, Attribute::NoCapture);
           }
            if (F->getArgument(4)->getType()->isPointerTy()) {
              F->addParamAttr(4, Attribute::NoCapture);
            }
```
would have been more clearer, but I might as well have asked about why we set the args in the derivcall init in the way they are set.
Since dot(x,y) == dot(y,x), but not axpy(a,x,y) = axpy(a,y,x) I assumed that replacing this code is only valid based on the types of dot if we have the guarantee that `call.getArgOperand(1)->getType()->isPointerTy()` implies `call.getArgOperand(3)->getType()->isPointerTy()` (and vice versa). However, I couldn't understand why we won't just have
```.cpp
            if (call.getArgOperand(1)->getType()->isPointerTy()) {
              F->addParamAttr(2, Attribute::ReadOnly);
              F->addParamAttr(2, Attribute::NoCapture);
              F->addParamAttr(4, Attribute::NoCapture);
            }
```
or (to keep with your daxpy based notation)
```.cpp
            if (F->getArgument(2)->getType()->isPointerTy()) {
              F->addParamAttr(2, Attribute::ReadOnly);
              F->addParamAttr(2, Attribute::NoCapture);
              F->addParamAttr(4, Attribute::NoCapture);
            }
```
Having two different conditions for setting ParamAttr 2 and 4 implied for me that it would be valid to mix and match those two calling conventions which you gave, which seemed odd to me. (because why would you have two different if's there otherwise?)
daxpy(int, double, double*, int, double*, int);
daxpy(int, double, int64, int, int64, int);

Now given that those two are not allowed to be mixed I'm much less confused and will just check for one of the arguments to see which declaration we use.
